# Task 2.1: Dataset Selection and Setup

**Paper**: *A Simple Geometric Interpretation of SVM using Stochastic Adversaries* — Livni, Crammer, Globerson (AISTATS 2012)

## Dataset Justification

We use `sklearn.datasets.load_digits`, which provides 1797 samples of 8×8 pixel handwritten digit images across 10 classes (digits 0–9) with 64 continuous features (pixel intensities in the range 0–16). This dataset is appropriate because the paper's core contribution is a novel multiclass SVM formulation — the (ℓ₂)∞SVM — and having 10 classes lets us observe meaningful differences between ℓ₂,∞ regularization and standard Frobenius regularization. The problem type (multiclass classification with continuous features) matches the experimental setting in the paper. However, the original paper evaluated on UCI datasets such as USPS (7291 training samples, 256 features) and Pendigits (7494 samples, 16 features); our digits dataset is smaller (1797 total) with lower dimensionality (64 features), so we expect somewhat different absolute performance. Despite this, the relative comparison between regularization strategies should still be informative.

In [1]:
# Set random seed for reproducibility
import numpy as np
np.random.seed(42)

import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

print("Random seed set to 42")
print("All imports successful")

Random seed set to 42
All imports successful


## Load the Dataset

We load the digits dataset from sklearn. This is a built-in dataset requiring no manual download steps.

In [2]:
# Load the digits dataset
digits = load_digits()
X_raw, y = digits.data, digits.target

print(f"Dataset shape: {X_raw.shape}")
print(f"Number of samples: {X_raw.shape[0]}")
print(f"Number of features: {X_raw.shape[1]}")
print(f"Number of classes: {len(np.unique(y))}")
print(f"Classes: {np.unique(y)}")
print(f"Samples per class: {dict(zip(*np.unique(y, return_counts=True)))}")
print(f"Raw feature range: [{X_raw.min():.1f}, {X_raw.max():.1f}]")

Dataset shape: (1797, 64)
Number of samples: 1797
Number of features: 64
Number of classes: 10
Classes: [0 1 2 3 4 5 6 7 8 9]
Samples per class: {np.int64(0): np.int64(178), np.int64(1): np.int64(182), np.int64(2): np.int64(177), np.int64(3): np.int64(183), np.int64(4): np.int64(181), np.int64(5): np.int64(182), np.int64(6): np.int64(181), np.int64(7): np.int64(179), np.int64(8): np.int64(174), np.int64(9): np.int64(180)}
Raw feature range: [0.0, 16.0]


## Standardize Features and Split Data

We standardize features to zero mean and unit variance using `StandardScaler`, as is standard practice for SVM methods. We split into 80% train / 20% test with `random_state=42` for reproducibility.

In [3]:
# Standardize features (zero mean, unit variance)
scaler = StandardScaler()
X = scaler.fit_transform(X_raw)

print(f"Standardized feature range: [{X.min():.4f}, {X.max():.4f}]")
print(f"Feature mean (should be ~0): {X.mean():.6f}")
print(f"Feature std (should be ~1): {X.std():.6f}")

Standardized feature range: [-3.0126, 42.3792]
Feature mean (should be ~0): 0.000000
Feature std (should be ~1): 0.976281


In [4]:
# Train/test split: 80/20, random_state=42
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

print(f"Training set: {X_train.shape[0]} samples")
print(f"Test set: {X_test.shape[0]} samples")
print(f"Train class distribution: {dict(zip(*np.unique(y_train, return_counts=True)))}")
print(f"Test class distribution: {dict(zip(*np.unique(y_test, return_counts=True)))}")

Training set: 1437 samples
Test set: 360 samples
Train class distribution: {np.int64(0): np.int64(142), np.int64(1): np.int64(146), np.int64(2): np.int64(142), np.int64(3): np.int64(146), np.int64(4): np.int64(145), np.int64(5): np.int64(145), np.int64(6): np.int64(145), np.int64(7): np.int64(143), np.int64(8): np.int64(139), np.int64(9): np.int64(144)}
Test class distribution: {np.int64(0): np.int64(36), np.int64(1): np.int64(36), np.int64(2): np.int64(35), np.int64(3): np.int64(37), np.int64(4): np.int64(36), np.int64(5): np.int64(37), np.int64(6): np.int64(36), np.int64(7): np.int64(36), np.int64(8): np.int64(35), np.int64(9): np.int64(36)}


## Visualize Sample Images

Below we display a grid of sample digit images from the dataset to confirm the data loaded correctly.

In [5]:
# Show sample images from the dataset
fig, axes = plt.subplots(2, 5, figsize=(12, 5))
fig.suptitle('Sample Digit Images from sklearn.datasets.load_digits', fontsize=14)

for i, ax in enumerate(axes.flat):
    # Show one sample per class
    idx = np.where(y == i)[0][0]
    ax.imshow(digits.images[idx], cmap='gray', interpolation='nearest')
    ax.set_title(f'Label: {y[idx]}', fontsize=11)
    ax.axis('off')

plt.tight_layout()
plt.savefig('results/task_2_1_sample_images.png', dpi=150, bbox_inches='tight')
plt.show()
print("Sample images saved to results/task_2_1_sample_images.png")

Sample images saved to results/task_2_1_sample_images.png


/var/folders/0m/80y5mc015y10780x_m98qlhm0000gn/T/ipykernel_98854/79422722.py:14: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## Summary

| Property | Value |
|----------|-------|
| Dataset | sklearn.datasets.load_digits |
| Total samples | 1797 |
| Features | 64 (8×8 pixel intensities) |
| Classes | 10 (digits 0–9) |
| Train samples | 1437 |
| Test samples | 360 |
| Standardization | Zero mean, unit variance |
| Random seed | 42 |

This data is ready for use in Task 2.2 (reproducing the (ℓ₂)∞SVM from the paper) and Task 2.3 (results comparison).